### PR TITLE
KEP 98: Promotion Phase for GitOps use cases

### PR DESCRIPTION
**Pull request by mowies**

### CODE REVIEW

File: text/0098-promotion-phase.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Promotion Phase for GitOps use cases
+
+Add a new lifecycle phase after post deployment dedicated to GitOps promotion tasks.
+
+**State: DRAFTING**
+
+## Motivation
+
+Since one of the main use cases for Keptn is enhancement of GitOps workflows, we should have bigger focus on that
+by introducing a dedicated stage after the post-deployment phase for GitOps promotion tasks.
```
Comment (thisthat) [inline text/0098-promotion-phase.md:10]: To emph that's the very last thing happening and not at workload level

```suggestion
by introducing a dedicated stage after the post Application deployment phase for GitOps promotion tasks.
```

File: text/0098-promotion-phase.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+phase.
+
+## Prior art and alternatives
+
+The alternative to this KEP is that users re-purpose the post-deployment phase for promotion, which would limit
+flexibility since users lose the option to not promote when post-deployment tasks fail.
+
+## Open questions
+
+For now, none.
```
Comment (thisthat) [inline text/0098-promotion-phase.md:38]: how to deal the case where the App get deployed but a Task in the promotion step fails. will be the overall deployment reported as failure or success? 🤔

Comment (mowies) [reply text/0098-promotion-phase.md:38]: i think we should have a config option for users to define whether keptn will mark this as failed or not. similar to the non-blocking behaviour described [here](https://github.com/keptn/lifecycle-toolkit/issues/1934)

Comment (mowies) [reply text/0098-promotion-phase.md]: i added some content about this

**Comment by mowies** [comment]
@thisthat pls re-review :)

File: text/0098-promotion-phase.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Promotion Phase for GitOps use cases
+
+Add a new lifecycle phase after post deployment dedicated to GitOps promotion tasks.
+
+**State: DRAFTING**
+
+## Motivation
+
+Since one of the main use cases for Keptn is enhancement of GitOps workflows, we should have bigger focus on that
```
Comment (thisthat) [inline text/0098-promotion-phase.md:9]: I'd suggest dropping the GitOps wording in this section. The promotion task could potentially promote by applying the manifests directly into another cluster 😅

Comment (mowies) [reply text/0098-promotion-phase.md]: fixed

**Comment by thschue** [comment]
This EP makes much sense and enables many workflows (especially regarding promotion tasks). Could it be that this partially deprecates this one: https://github.com/keptn/enhancement-proposals/pull/98 (on-success tasks but implemented in a new phase)?

**Comment by mowies** [comment]
@thschue it's connected to the other KEP a bit, yes. But only in part of course. The thing is that you can't reference tasks here. So you basically have an unordered list of  tasks that are run as a result of the success of the post-deployment phase.
But there's no way of referencing specific tasks when a specific post-deployment task succeeded for example.

**Review by thschue — APPROVED** [review]
lgtm

**Review by thisthat — APPROVED** [review]

**Comment by heckelmann** [comment]
Two questions: Will this phase being executed as well when the status of a post deployment task or evaluation is failing? If so, do I get the results in my promotion task as well?

**Comment by mowies** [comment]
No, it will only execute if the phases before it succeeded.

**Comment by mowies** [comment]
epic ticket for this: https://github.com/keptn/lifecycle-toolkit/issues/2907

**Pull request merged**